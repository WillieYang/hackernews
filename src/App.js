import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';

const list = [
  {
    title: 'React',
    url: 'https://facebook.github.io/react/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://github.com/react.js/redux',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  }
];
const searchTerm = '';

const isSearched = searchTerm => item =>
  item.title.toLowerCase().includes(searchTerm.toLowerCase());

class App extends Component {
  constructor(props) {
    super(props);
    this.state = { list, searchTerm };
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
  }
  
  onDismiss(id) {
    const updatedList = this.state.list.filter(item => item.objectID !== id);
    this.setState({ list: updatedList }); 
  }
  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value })
  }
  
  render() {
    const helloWorld = "Welcome to the World of React";
    const username = {
        firstName: 'Sheng',
        lastName: 'Yang',
    };
    const { list, searchTerm } = this.state;
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
        </header>

        <h2>{ helloWorld }</h2>
        <p>{ username.firstName } { username.lastName }</p>
        <Search
          value={ searchTerm }
          onChange={this.onSearchChange}
        > Search
        </Search>
        <Table
          list={ list }
          pattern={ searchTerm }
          onDismiss={ this.onDismiss }
        />
        {/*<form>*/}
          {/*<input type="text"*/}
                 {/*onChange={this.onSearchChange} />*/}
        {/*</form>*/}
        {/*{*/}
          {/*list.filter(isSearched(searchTerm)).map(item =>*/}
              {/*<div key={ item.objectID }>*/}
                {/*<span>*/}
                  {/*<a href={ item.url }>{ item.title }</a>*/}
                {/*</span>*/}
                {/*<span>{ item.author }</span>*/}
                {/*<span>{ item.num_comments }</span>*/}
                {/*<span>{ item.points }</span>*/}
                {/*<span>*/}
                  {/*<button*/}
                    {/*onClick={() => this.onDismiss(item.objectID)}*/}
                    {/*type="button"*/}
                  {/*>*/}
                    {/*Dismiss*/}
                  {/*</button>*/}
                {/*</span>*/}
              {/*</div>*/}
          {/*)}*/}
      </div>
    );
  }
}

class Search extends  Component {
  render() {
    const { value, onChange, children } = this.props;
    return (
      <form>
        { children } <input
          type="text"
          value={value}
          onChange={onChange}
        />
      </form>
      );
  }
}

class Table extends Component {
  render() {
    const { list, pattern, onDismiss } = this.props;
    return (
      <div>
        { list.filter(isSearched(pattern)).map(item =>
          <div key={item.objectID}>
            <span>
              <a href={ item.url }>{ item.title }</a>
            </span>
            <span>{ item.author }</span>
            <span>{ item.num_comments }</span>
            <span>{ item.points }</span>
            <span>
              <button
                onClick={() => onDismiss(item.objectID)}
                type="button"
              >
                Dismiss
              </button>
            </span>
          </div>
        )}
      </div>
    );
  }
}

export default App;
