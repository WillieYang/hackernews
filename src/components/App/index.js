import React, { Component } from 'react';
import { connect } from "react-redux";

// import { startAction } from "../../actions/startAction";
// import { stopAction } from "../../actions/stopAction";
import rotateAction from '../../actions/rotateAction'
import logo from '../../logo.svg';
import './index.css';
import Search from '../Search'
import Button from '../Button'
import Table from '../Table'
import Loading from '../Loading'
import {
  PARAM_HPP,
  DEFAULT_HPP,
  DEFAULT_QUERY,
  PARAM_PAGE,
  PARAM_SEARCH,
  PATH_BASE,
  PATH_SEARCH } from '../../constants'

import fetch from 'isomorphic-fetch';
import { library } from '@fortawesome/fontawesome-svg-core'
import { faStroopwafel, faSpinner } from '@fortawesome/free-solid-svg-icons'

library.add(faStroopwafel, faSpinner);

const UpdateSearchTopStoriesState = (hits, page) => (prevState) => {
  const { searchKey, results } = prevState;

  const oldHits = results && results[searchKey]
    ? results[searchKey].hits
    : [];

  const updateHits = [
    ...oldHits,
    ...hits
  ];

  return {
    results: {
      ...results,
      [searchKey]: { hits: updateHits, page}
    },
    isLoading: false
  };
};

// const url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}&${PARAM_PAGE}`;

// const isSearched = searchTerm => item =>
//   item.title.toLowerCase().includes(searchTerm.toLowerCase());

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY,
      error: null,
      isLoading: false,
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
  }

  needsToSearchTopStories(searchTerm) {
    return !this.state.results[searchTerm];
  }

  setSearchTopStories(result) {
    const { hits, page} = result;

    this.setState(UpdateSearchTopStoriesState(hits, page));
  }

  fetchSearchTopStories(searchTerm, page = 0) {
    this.setState({ isLoading: true });

    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(response => response.json())
      .then(result => this.setSearchTopStories(result))
      .catch(e => this.setState({ error: e }));
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm});
    this.fetchSearchTopStories(searchTerm)
  }

  onDismiss(id) {
    this.setState(prevState => {
      const {searchKey, results} = prevState;
      const {hits, page} = results[searchKey];
      const isNotId = item => item.objectID !== id;
      const updatedList = hits.filter(isNotId);
      return {
        results: {
        ...results,
            [searchKey]: {hits: updatedList, page}
        }
      };
    });
  }

  onSearchChange(event) {
    this.setState({ searchTerm: event.target.value })
  }

  onSearchSubmit(event) {
    const { searchTerm } = this.state;
    this.setState({ searchKey: searchTerm });

    if(this.needsToSearchTopStories(searchTerm)){
      this.fetchSearchTopStories(searchTerm)
    }
    event.preventDefault();
  }
  
  render() {
    const helloWorld = "Welcome to the World of React";
    const username = {
        firstName: 'Sheng',
        lastName: 'Yang',
    };
    const { results, searchTerm, searchKey, error, isLoading } = this.state;
    const page = (results && results[searchKey] && results[searchKey].page) || 0;
    const list = (results && results[searchKey] && results[searchKey].hits) || [];

    return (
      <div className="App">
        <header className="App-header">
          <img src={logo}
               className={
                 "App-logo" + (this.props.rotating ? "": " App-logo-paused ")
               }
               alt="logo"
               onClick={
                 () => this.props.rotateAction(!this.props.rotating)
               }
          />
        </header>
        <h2>{ helloWorld }</h2>
        <p>{ username.firstName } { username.lastName }</p>
        <div className="interactions">
          <Search
            value={ searchTerm }
            onChange={this.onSearchChange}
            onSubmit={this.onSearchSubmit}>
            Search
          </Search>
        </div>
        { error
          ? <div className="interactions">
            <p>Something went wrong.</p>
          </div>
          : <Table
            list={ list }
            onDismiss={ this.onDismiss }
          />
        }
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
            More
          </ButtonWithLoading>
        </div>
      </div>
    );
  }
}

const withLoading = (Component) => ({ isLoading, ...rest }) =>
  isLoading
    ? <Loading />
    : <Component {...rest} />

const ButtonWithLoading = withLoading(Button);

const mapStateToProps = state => ({
  ...state
});

const mapDispatchToProps = dispatch => ({
  rotateAction: (payload) => dispatch(rotateAction(payload))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);