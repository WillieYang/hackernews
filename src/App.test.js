import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App, { Search, Button, Table } from './App';

describe('App', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <App />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Search', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<search>Search</search>, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <search>Search</search>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Button', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Button>Give me More</Button>, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Button>Give me More</Button>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y'},
      { title: '2', author: '2', num_comments: 1, points: 2, objectID: 'z'},
    ],
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('has a valid snapshot', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});