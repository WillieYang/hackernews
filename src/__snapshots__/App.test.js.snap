// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has a valid snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
  </header>
  <h2>
    Welcome to the World of React
  </h2>
  <p>
    Sheng
     
    Yang
  </p>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
         Search
      </button>
    </form>
  </div>
  <div
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
    >
      More
    </button>
  </div>
</div>
`;
